### ISSUE TITLE
S2-03 [Backend] FinBERT sentiment analyzer — ProsusAI/finbert (pre-trained, zero training)

### ISSUE THREAD

**Issue by NathanLuckock** [Nathan7108/HackathonPractice#14]
## Problem
The feature pipeline outputs 7 sentiment features but they're all zeros/defaults without a real NLP model. Headlines about military strikes, sanctions, and ceasefires need to be scored as negative/positive/neutral to feed the risk scorer. Without sentiment, the model is blind to media signals.

## Goal
Implement the FinBERT sentiment analyzer using ProsusAI/finbert (pre-trained, zero training). Load once at startup, batch-analyze headlines, return 7 sentiment features + individual headline results.

### In scope
- load_finbert() — downloads and caches ProsusAI/finbert (~440MB first run)
- analyze_headlines_sentiment(headlines: list[str]) -> dict with 7 features + individual_results
- Module-level singleton (load once, reuse across calls)
- GPU auto-detection (CUDA if available, CPU fallback)
- Batch processing in groups of 16

### Out of scope
- Fetching headlines from NewsAPI (separate issue)
- Fine-tuning or retraining (using pre-trained model as-is)
- Integration with FastAPI (Sprint 3)

## Acceptance Criteria
- [ ] load_finbert() downloads model and returns pipeline object
- [ ] analyze_headlines_sentiment([]) returns EMPTY_SENTIMENT defaults (neutral)
- [ ] analyze_headlines_sentiment(5 conflict headlines) returns dominant_sentiment='negative'
- [ ] analyze_headlines_sentiment(["Ceasefire agreement signed"]) returns dominant_sentiment='positive'
- [ ] Returns exactly these 8 keys: finbert_negative_score, finbert_positive_score, finbert_neutral_score, headline_volume, headline_escalatory_pct, media_negativity_index, sentiment_trend_7d, dominant_sentiment
- [ ] individual_results list contains {headline, label, score} per headline
- [ ] Batch size 16 for efficiency
- [ ] Running python -m backend.ml.sentiment prints test results

### ProsusAI/finbert labels:
- negative = escalatory (sanctions, conflict, crisis, attack, collapse)
- positive = de-escalatory (ceasefire, agreement, talks, aid, recovery)
- neutral = no strong signal

### Implementation (from ML Guide Section 3.3):
```python
from transformers import pipeline
import torch
import numpy as np

_finbert_pipeline = None

def load_finbert():
    global _finbert_pipeline
    if _finbert_pipeline is None:
        print("Loading ProsusAI/finbert...")
        _finbert_pipeline = pipeline(
            "sentiment-analysis",
            model="ProsusAI/finbert",
            tokenizer="ProsusAI/finbert",
            device=0 if torch.cuda.is_available() else -1,
            max_length=512,
            truncation=True,
        )
        print("finbert loaded successfully")
    return _finbert_pipeline

def analyze_headlines_sentiment(headlines: list[str]) -> dict:
    if not headlines:
        return {
            'finbert_negative_score': 0.0,
            'finbert_positive_score': 0.0,
            'finbert_neutral_score': 1.0,
            'headline_escalatory_pct': 0.0,
            'headline_volume': 0,
            'media_negativity_index': 0.0,
            'sentiment_trend_7d': 0.0,
            'dominant_sentiment': 'neutral',
        }
    pipe = load_finbert()
    batch_size = 16
    all_results = []
    for i in range(0, len(headlines), batch_size):
        batch = headlines[i:i+batch_size]
        results = pipe(batch)
        all_results.extend(results)
    # Aggregate scores
    neg_scores = [r['score'] for r in all_results if r['label'] == 'negative']
    pos_scores = [r['score'] for r in all_results if r['label'] == 'positive']
    neu_scores = [r['score'] for r in all_results if r['label'] == 'neutral']
    neg_pct = len(neg_scores) / max(len(all_results), 1)
    avg_neg = float(np.mean(neg_scores)) if neg_scores else 0.0
    avg_pos = float(np.mean(pos_scores)) if pos_scores else 0.0
    counts = {'negative': len(neg_scores), 'positive': len(pos_scores), 'neutral': len(neu_scores)}
    dominant = max(counts, key=counts.get)
    return {
        'finbert_negative_score': round(avg_neg, 3),
        'finbert_positive_score': round(avg_pos, 3),
        'finbert_neutral_score': round(1 - avg_neg - avg_pos, 3),
        'headline_escalatory_pct': round(neg_pct, 3),
        'headline_volume': len(headlines),
        'media_negativity_index': round(avg_neg * neg_pct, 3),
        'sentiment_trend_7d': 0.0,
        'dominant_sentiment': dominant,
        'individual_results': [
            {'headline': h[:80], 'label': r['label'], 'score': round(r['score'], 3)}
            for h, r in zip(headlines, all_results)
        ]
    }
```

### Test headlines (use these in __main__):
```python
test_headlines = [
    "Russia launches missile strikes on Kyiv infrastructure",
    "Iran nuclear talks collapse as IAEA inspectors expelled",
    "Taiwan Strait military exercises draw US Navy response",
    "Venezuela opposition leader arrested ahead of elections",
    "Ceasefire agreement signed between Ethiopia and rebels",
    "Pakistan military launches offensive in tribal regions",
    "Brazil economic growth exceeds forecasts at 3.2%",
    "Serbia moves troops near Kosovo border amid tensions",
]
```

### Edge cases:
- Empty headlines list: return neutral defaults
- Very long headlines: truncation=True handles >512 tokens
- Non-English headlines: finbert is English-only, may produce unexpected labels. Acceptable for hackathon.
- First call downloads ~440MB: subsequent calls use cached model from ~/.cache/huggingface/

### Hackathon tradeoff:
- sentiment_trend_7d always 0.0 (would need 7-day rolling history)

## Manual Test Steps
1. python -m backend.ml.sentiment
2. Verify "Russia launches missile strikes" -> negative with high score
3. Verify "Ceasefire agreement signed" -> positive
4. Verify "Brazil economic growth exceeds forecasts" -> positive
5. Verify dominant_sentiment is 'negative' (5 of 8 headlines are escalatory)
6. Verify headline_escalatory_pct > 0.5

---

## Cursor Agent Instructions

Repo: Nathan7108/HackathonPractice
Working directory: C:\Users\natha\HackathonPractice
Branch: feature/s2-03-finbert-sentiment

### STEP 0 — Read context
```
cat PROGRESS.md
```

### STEP 1 — Create branch
```
git checkout main && git pull origin main
git checkout -b feature/s2-03-finbert-sentiment
```

### STEP 2 — Implement backend/ml/sentiment.py
Replace placeholder with full implementation per this issue. Include:
1. Module-level _finbert_pipeline singleton
2. load_finbert() with GPU auto-detection
3. analyze_headlines_sentiment() with batch processing and all 8 return keys
4. __main__ block with 8 test headlines, print results

### STEP 3 — Validate
```
python -m backend.ml.sentiment
```
Verify: model downloads, test headlines scored, dominant_sentiment printed.

### STEP 4 — Commit and push
```
git add backend/ml/sentiment.py
git commit -m "feat: S2-03 FinBERT sentiment analyzer via ProsusAI/finbert"
git push -u origin feature/s2-03-finbert-sentiment
```

### STEP 5 — Open PR
```
gh pr create --title "S2-03 [Backend] FinBERT sentiment analyzer" --body "Closes #12" --base main
```

### Definition of Done
- [ ] load_finbert() works (downloads model on first run)
- [ ] analyze_headlines_sentiment() returns correct 8 keys
- [ ] Negative headlines score as negative
- [ ] Positive headlines score as positive
- [ ] Empty list returns neutral defaults
- [ ] Pushed to origin, PR created

### If You Get Stuck
- Model name: ProsusAI/finbert (not just finbert)
- If GPU errors: set device=-1 to force CPU
- If import errors: pip install transformers torch
- Model cache: ~/.cache/huggingface/hub/
